lib/elixir/lib/calendar/iso.ex

```diff
@@ -258,4 +384,21 @@ defmodule Calendar.ISO do
       _ -> :error
     end
   end
+
+  def valid_date?(year, month, day)
+  def valid_date?(year, month, day) do
+    :calendar.valid_date(year, month, day) and year <= 9999
+  end
+  def valid_date?(year, month, day), do: false
+
+  def valid_time?(hour, minute, second, microsecond)
+  def valid_time?(hour, minute, second, {microsecond, _})
+  when hour in 0..23 and minute in 0..59 and second in 0..60 and microsecond in 0..999_999 do
```

Comment:
Remove the guard and have this check inline and you will be able to make this one clause instead of three.